### PR TITLE
PineTab2 UCM2 profile

### PR DESCRIPTION
**Pull request by ScottFreeCode**
This solves the problem of the speakers not turning off when the headphone jack is plugged in.

There are some oddities found and worked around, noted in the code.

Code is based on the UCM conf's rk817 profile, with modifications to work around a couple PineTab2 issues. The main outstanding issues are documented in comments next to the workarounds but I also did hardcode the playback mux to HP removing the SPK option as SPK seemed to disable sound in practice.

Packaging is based on alsa-ucm-pinetab.

### CODE REVIEW

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Maintainer: Danct12 <danct12@disroot.org>
```
Comment (ScottFreeCode) [inline PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD:1]: Let me know if you'd prefer I maintained this; but I presume it should be handed off as just another PineTab2 package/configuration!

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Syntax 4
+
+SectionUseCase."HiFi" {
+	File "/Pine64/PineTab/HiFi.conf"
```
Comment (ScottFreeCode) [inline PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf:4]: I originally put these under Rockchip rather than Pine64, like the configuration I based them on. However, I switched following the PineTab1's structure when I made this PR.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+pkgname=alsa-ucm-pinetab2
+pkgver=0.1.0
+pkgrel=1
+pkgdesc="UCM files for PineTab 2"
+arch=(any)
+url="https://github.com/dreemurrs-embedded/Pine64-Arch"
+license=('MIT')
```
Comment (ScottFreeCode) [inline PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD:8]: I _think_ the license of the package containing the rk817 profile I based this upon is BSD. I recall both of these are permissive licenses. That said, I forget if either requires attribution or any other special considerations. Let me know if we need to change this from MIT or if there are any other issues.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf [hunk 4]
```diff
@@ -0,0 +1,74 @@
+Define.pbk_mux "SPK"
```
Comment (ScottFreeCode) [inline PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:1]: One more I almost forgot… this is left over from the rk8whatever it was. I don't know what SPK in the Mux means or is supposed to do but it seemed to make the PineTab2 stop producing sound, so, I hardcoded it to HP instead.

Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:1]: (We could put more of this info, "based on BSD-licensed rk817" and "don't use this if you want sound" or whatever, in comments in the code for posterity?)

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf (hunk 2 above)
Comment (Danct12) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf:4]: Directory name should be PineTab2, otherwise there'll be conflicts.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD (hunk 3 above)
Comment (Danct12) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD:8]: Change it to BSD.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD (hunk 1 above)
Comment (Danct12) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PKGBUILD:1]: Since this is your package that you submitted, you can maintain it.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf (hunk 2 above)
Comment (Danct12) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf:4]: Then that's a kernel issue. CC: @TuxThePenguin0

Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf]: Went ahead and updated everything to "PineTab2" on the assumption that it will work once there's a change to make the chip be named "PineTab2" in the system. Easy enough to change back or change to whatever else if we go with a different solution.

Comment (TuxThePenguin0) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/PineTab.conf]: should be fixed to use PineTab2 now

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		cset "name='Speakers Switch' on"
+	]
+
+	Value {
+		PlaybackMixerElem "Master Playback Volume"
+		PlaybackPriority 100
+		PlaybackPCM "hw:${CardId}"
+		# Headphones plugged/unplugged detection is backwards on PT2.
+		# The system sees them as "unplugged" when they are plugged in
+		# and not when they are actually physically unplugged!
```
[the remark this answers is not in the document]
Comment (Danct12) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:35]: This is also another kernel issue. Easy fix by modifying the device tree.

```
diff --git a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
index a766f21bd6f8..1ba6e610767a 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
@@ -135,7 +135,7 @@ rk817-sound {
                        "Speakers", "Speaker Amplifier OUTL",
                        "Speakers", "Speaker Amplifier OUTR";
 
-               simple-audio-card,hp-det-gpio = <&gpio4 RK_PC6 GPIO_ACTIVE_HIGH>;
+               simple-audio-card,hp-det-gpio = <&gpio4 RK_PC6 GPIO_ACTIVE_LOW>;
                simple-audio-card,aux-devs = <&speaker_amp>;
                simple-audio-card,pin-switches = "Speakers", "Microphone";

```

After applying the patch, recompile the device tree and boot.

Plugged:
```
Event: time 1688759421.478541, type 5 (EV_SW), code 2 (SW_HEADPHONE_INSERT), value 1
Event: time 1688759421.478541, -------------- SYN_REPORT ------------
```

Unplugged:
```
Event: time 1688759449.618550, type 5 (EV_SW), code 2 (SW_HEADPHONE_INSERT), value 0
Event: time 1688759449.618550, -------------- SYN_REPORT ------------
```

CC: @TuxThePenguin0

[the remark this answers is not in the document]
Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:35]: Okay; if we fix that in the device tree, we'll want to move `JackControl "Headphones Jack"` back to the  headphones section where it is in Rockchip/rk817_sound (and remove these comments from next to where it is now).

[the remark this answers is not in the document]
Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:35]: Actually – If you've got a pair of headphones handy, could you try…

1. Make the device tree fix for the backwards plugged state.
2. Change the soundchip name (the one we already corrected in the kernel [or device tree?] from "PineTab" to "PineTab2") to "rk817_int" or "rk817_ext" (I'm not sure why there are these options under conf.d/simple-card, *both* point to "Rockchip/rk817-sound").
3. Rebuild the kernel and boot into the new kernel.
4. Remove this Pinetab2 profile locally.
5. Try plugging and unplugging headphones, turning up the volume till you can hear sound clearly in both states (since they have separate volume settings), and see if the results are:
  - With them plugged, sound comes through the headphones.
  - With them plugged, sound does _not_ come through the speaker.
  - With them unplugged, sound comes through the speaker again.
  - All of the above hold after plugging and unplugging at least once before (i.e. no weirdness where after the first plug/unplug one mode or the other stops working; I ran into that while testing this).

**I'm wondering if the existing profile for this chip/board would have worked if we'd not had the plug/unplug state backwards.** It's possible any other problems I encountered and had to work around, were downstream of that or of my initial attempts to fix things.

**Comment by ScottFreeCode** [comment]
> Wouldn't it be better if this was submitted directly to the upstream repo? https://github.com/alsa-project/alsa-ucm-conf/
> 
> See here for the PinePhone's submission: https://github.com/alsa-project/alsa-ucm-conf/pull/134

In the long run, yes, upstreaming any needed new profile would be ideal. But the PineTab (1) is still in this repo and not that one, too. And we're still working out the bugs (see e.g. the conversation about the backwards plugged/unplugged state). And after we work out all the bugs it might turn out we don't need this since there's an existing Rockchip profile upstream.

File: PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+If.1 {
+	Condition {
+		Type ControlExists
+		Control "name='Internal Speakers Switch'"
+	}
+
+	True {
+		Define.pbk_mux "HP"
+		SectionDevice."Speaker".EnableSequence [
+			cset "name='Internal Speakers Switch' on"
```
Comment (ScottFreeCode) [inline PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:12]: (TL/DR list at the bottom.)

Hey, I was thinking about the mux setting that's in the upstream rk817-sound profile but not in this one. It defaults to SPK or something that doesn't work on the PineTab2, I had to force it to HP which works on the PT2. But! The conditional section sets it to HP if there is an "Internal Speakers Switch".

Get this: The same section would also add commands that do the same as my hardcoded speaker on/off commands, but for "Internal Speakers Switch" instead of my "Internal Speakers" setting!

@TuxThePenguin0 Do you know what would have to be changed in the device tree to rename the "Speakers Switch" to "Internal speakers Switch"? I don't see a "Speakers Switch" in there directly, it looks like it's generated from a list of switches but I suspect other things named "Speakers" have to match the `s/Speakers/Internal Speakers/` change too. If I end up having time to learn to recompile the device tree I will play around with it but if you already know what should and shouldn't be changed to rename that, you might be able to make the fix it quicker than I can.

TL/DR: My guess would be if we make three changes/fixes:

- @Danct12 's fix for the plugged state
- rename "Speakers Switch" to "Internal Speakers Switch" (however that's done, exactly)
- rename the sound card from "PineTab2" to "rk817_int"

Then it probably almost certainly should work with the upstream sound profile.

Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:12]: [Corrected mistype in second paragraph.]

Comment (diederikdehaas) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:12]: > rename "Speakers Switch" to "Internal Speakers Switch" (however that's done, exactly)

https://www.kernel.org/doc/Documentation/devicetree/bindings/sound/widgets.txt makes me think it should be changed in this part of the DT patch:
```patch
+		simple-audio-card,widgets =
+			"Microphone", "Mic Jack",
+			"Headphone", "Headphones",
+			"Microphone", "Microphone",
+			"Speaker", "Speakers";
```

(not so sure whether it's useful or a good idea though)

Comment (ScottFreeCode) [reply PKGBUILDS/pine64/alsa-ucm-pinetab2/HiFi.conf:12]: That looks like it's for the control panel, not sure if it does or doesn't have to match the "Speakers" elsewhere in the device tree. There's another line that's something like `switches = "Microphone", "Speakers"` which I'm fairly sure we'll need to change if this is the right approach.

**Comment by ScottFreeCode** [comment]
@Danct12 @TuxThePenguin0 I have built and tested a fix to the device tree to get it to work with the upstream rk817-sound profile, and it works perfectly! (I removed my PineTab2 profile before testing it, so I know it is using the upstream one.)

This is the patch if you HAVE NOT applied [@Danct12's patch upthread for the gpio high/low](https://github.com/dreemurrs-embedded/Pine64-Arch/pull/555#discussion_r1256390498):

```patch
diff --git a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
index a766f21bd..60d28ba24 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
@@ -116,14 +116,14 @@ rk817-sound {
 		pinctrl-names = "default";
 		pinctrl-0 = <&hp_det_l>;
 		simple-audio-card,format = "i2s";
-		simple-audio-card,name = "PineTab2";
+		simple-audio-card,name = "rk817_int";
 		simple-audio-card,mclk-fs = <256>;
 
 		simple-audio-card,widgets =
 			"Microphone", "Mic Jack",
 			"Headphone", "Headphones",
 			"Microphone", "Microphone",
-			"Speaker", "Speakers";
+			"Speaker", "Internal Speakers";
 
 		simple-audio-card,routing =
 			"MICL", "Microphone",
@@ -132,12 +132,12 @@ rk817-sound {
 			"Headphones", "HPOR",
 			"Speaker Amplifier INL", "HPOL",
 			"Speaker Amplifier INR", "HPOR",
-			"Speakers", "Speaker Amplifier OUTL",
-			"Speakers", "Speaker Amplifier OUTR";
+			"Internal Speakers", "Speaker Amplifier OUTL",
+			"Internal Speakers", "Speaker Amplifier OUTR";
 
-		simple-audio-card,hp-det-gpio = <&gpio4 RK_PC6 GPIO_ACTIVE_HIGH>;
+		simple-audio-card,hp-det-gpio = <&gpio4 RK_PC6 GPIO_ACTIVE_LOW>;
 		simple-audio-card,aux-devs = <&speaker_amp>;
-		simple-audio-card,pin-switches = "Speakers", "Microphone";
+		simple-audio-card,pin-switches = "Internal Speakers", "Microphone";
 
 		simple-audio-card,cpu {
 			sound-dai = <&i2s1_8ch>;
```

This is the patch if you HAVE applied [@Danct12's patch upthread for the gpio high/low](https://github.com/dreemurrs-embedded/Pine64-Arch/pull/555#discussion_r1256390498):

```patch
diff --git a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
index 1ba6e6107..60d28ba24 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3566-pinetab2.dtsi
@@ -116,14 +116,14 @@ rk817-sound {
 		pinctrl-names = "default";
 		pinctrl-0 = <&hp_det_l>;
 		simple-audio-card,format = "i2s";
-		simple-audio-card,name = "PineTab2";
+		simple-audio-card,name = "rk817_int";
 		simple-audio-card,mclk-fs = <256>;
 
 		simple-audio-card,widgets =
 			"Microphone", "Mic Jack",
 			"Headphone", "Headphones",
 			"Microphone", "Microphone",
-			"Speaker", "Speakers";
+			"Speaker", "Internal Speakers";
 
 		simple-audio-card,routing =
 			"MICL", "Microphone",
@@ -132,12 +132,12 @@ rk817-sound {
 			"Headphones", "HPOR",
 			"Speaker Amplifier INL", "HPOL",
 			"Speaker Amplifier INR", "HPOR",
-			"Speakers", "Speaker Amplifier OUTL",
-			"Speakers", "Speaker Amplifier OUTR";
+			"Internal Speakers", "Speaker Amplifier OUTL",
+			"Internal Speakers", "Speaker Amplifier OUTR";
 
 		simple-audio-card,hp-det-gpio = <&gpio4 RK_PC6 GPIO_ACTIVE_LOW>;
 		simple-audio-card,aux-devs = <&speaker_amp>;
-		simple-audio-card,pin-switches = "Speakers", "Microphone";
+		simple-audio-card,pin-switches = "Internal Speakers", "Microphone";
 
 		simple-audio-card,cpu {
 			sound-dai = <&i2s1_8ch>;
```

If there isn't any reason that the device tree _needs_ to name that "Speakers" (as it is now) instead of "Internal Speakers" (as I'm proposing changing it to), or that we can't/shouldn't reuse the existing rk817_(int|ext) profile/link, then I would highly recommend closing out this PR and skipping this package in favor of using this for the upstream rk817-sound profile!

**Comment by diederikdehaas** [comment]
I can confirm that **just** the DeviceTree changes suggested, makes audio work properly.

**Comment by ScottFreeCode** [comment]
Never before have I been so happy to have a pull request of mine _not_ be merged! Looking forward to this making it out of the -testing package repo.

**Comment by diederikdehaas** [comment]
Any particular reason you went with ``rk817_ext`` and not ``rk817_int``?

I have no idea what the reason is/was for both existing (upstream) and one seems to be a symlink to the other, so AFAIK it doesn't make a difference. But I'm still curious why you went with *_ext* and not *_int* as proposed in this MR.

**Comment by Danct12** [comment]
The PineTab 2 has external amplifier setup, so I went with `rk817_ext` because that's what they do too: https://github.com/alsa-project/alsa-ucm-conf/commit/b0497ca4b508d6f894d78b16e0e06616a2a36c16

**Comment by ScottFreeCode** [comment]
Ah, good find. I had assumed "int" because he had to specify "Internal Speakers" to get it to work.
